### DISCUSSION TITLE
1.1.373 issue - Expected class but received "Literal[0]" & Too few type arguments provided for "FixtureFunction"

### DISCUSSION THREAD

**Discussion by lexicalunit** [microsoft/pyright#8608] [General]
**Describe the bug**
In version `1.1.373` I'm getting a bunch of cryptic errors with code that passed fine in version `1.1.372`.

**Code or Screenshots**

I was able to create a minimal example:

```python
from sqlalchemy.engine import create_engine
from sqlalchemy.orm import sessionmaker

database_url = "whatever:///does/not/matter"
engine = create_engine(database_url)
connection = engine.connect()
session_maker = sessionmaker(bind=connection)
session = session_maker()
rows = session.execute("SELECT xid FROM guilds").fetchall()
xids = [int(row[0]) for row in rows]
```

You can see the full error logs in my github actions workflow: https://github.com/lexicalunit/spellbot/actions/runs/10168666850/job/28123798890

**VS Code extension or command-line**
- [x] Command Line

```sh
➜ poetry run pyright tests src example.py
/path/to/spellbot/example.py
  /path/to/spellbot/example.py:10:17 - error: Expected class but received "Literal[0]" (reportGeneralTypeIssues)
  /path/to/spellbot/example.py:10:17 - error: Too few type arguments provided for "FixtureFunction[SimpleFixtureFunction@FixtureFunction, FactoryFixtureFunction@FixtureFunction]"; expected 2 but received 1 (reportInvalidTypeForm)
```

However, if I run `pyright` on a single file, I get no error:

```sh
➜ poetry run pyright example.py
0 errors, 0 warnings, 0 informations
```

It only happens when I execute `pyright` on all my files.

**Additional Information**

```sh
➜ poetry run pip freeze | grep -iE "sqlalchemy|pyright"
pyright==1.1.373
SQLAlchemy==2.0.31
sqlalchemy-stubs==0.4
SQLAlchemy-Utils==0.41.2

➜ poetry run python --version
Python 3.12.4
```

**Comment by erictraut** [comment]
I'm not able to repro what you're seeing. If I install the "sqlalchemy-stubs` package, I see no errors from pyright. I'm guessing that you have something else installed in your venv that's affecting the results. Are you able to repro the problem with a clean venv?

My understanding is that sqlalchemy 2.x is fully typed — that is, it has inline type information — and that it doesn't require the use of an external stub library. By adding the stub library, you're overriding the type information provided by the library maintainers with outdated and incomplete type information. I think the stub library is intended for use with sqlalchemy 1.x, which didn't have type information. You might want to try removing it from your environment. If I don't install `sqlalchemy-stubs`, I receive a type error for the `session.execute` call because this call apparently doesn't accept a `str` argument.

**Reply by lexicalunit** [reply]
I get the same errors listed above if I uninstall `sqlalchemy-stubs`. However removing these stubs causes additional errors like, for example:
```log
test_games.py:338:24 - error: Argument of type "Column[int]" cannot be assigned to parameter "author_xid" of type "int"
    "Column[int]" is incompatible with "int" (reportArgumentType)
```

**Reply by lexicalunit** [reply]
Weirdly, I don't get the same type errors in the editor with `pylance` that I do from `pyright`. Not sure what that's about either.

**Reply by erictraut** [reply]
Pylance is typically using a version of pyright that's one to four weeks behind. You could try switching to the prerelease version of pylance.

**Comment by erictraut** [comment]
Moving to a discussion topic because I don't think this is a bug in pyright.

**Reply by lexicalunit** [reply]
Reverting to `1.1.372` seems to fix the issue for now at least.

**Comment by kjagiello** [comment]
Googling for `Too few type arguments provided for "FixtureFunction[SimpleFixtureFunction@FixtureFunction, FactoryFixtureFunction@FixtureFunction]"...` lead me to this discussion.

I have bumped pyright in a project and got a multitude of these as well. The issue seems to stem from the pytest-asyncio package. For some reason, as soon as `pytest_asyncio.fixture` appears in the code, all the untyped arguments become `FixtureFunction[SimpleFixtureFunction@FixtureFunction, FactoryFixtureFunction@FixtureFunction]`.

I think this can serve as a minimal repro case:

```python
import pytest_asyncio

pytest_asyncio.fixture

def f(x):
    x["y"]
```

Output:
```
repro.py:6:7 - error: Too few type arguments provided for "FixtureFunction[SimpleFixtureFunction@FixtureFunction, FactoryFixtureFunction@FixtureFunction]"; expected 2 but received 1 (reportInvalidTypeForm)
repro.py:6:8 - error: "y" is not defined (reportUndefinedVariable)
```

Packages:
```
pyright==1.1.374
pytest==8.3.2
pytest-asyncio==0.23.8
```

It seems like this issue appeared in 1.1.373.

**Reply by kjagiello** [reply]
An attempt to create a repro case without any external dependencies:

```python
from collections.abc import Callable
from typing import TypeAlias, TypeVar

A = TypeVar("A")
B = TypeVar("B")
C: TypeAlias = A | B
D: TypeAlias = Callable[[], C]

def f(x):
    x["y"]
```

([playground](https://pyright-play.net/?reportMissingModuleSource=true&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGEDFjkAoUSKVATwvSwHMylarQAqPJAEF46AjgA0UcRSQA1AiBYtJUALxKJ6kAAoARJNMBKFgCE9BlUbM2rLOgC4HUmXPs6APlA2LAAinsresjj2DExEyADaeMmKdAC6WgAmSMBQwMYAHpbuLFBlUAUJplymaUA))

The union alias seems to be important. For example this does not exhibit the same issue:
```python
from collections.abc import Callable
from typing import TypeAlias, TypeVar

A = TypeVar("A")
B = TypeVar("B")
C: TypeAlias = Callable[[], A | B]

def f(x):
    x["y"]
```

**Reply by erictraut** [reply]
Thanks for the great repro steps! This turns out to be a very bad bug in pyright. It has existed for a long time — at least a couple of years. It results in internal data structure corruption which results in very strange behaviors. I'm really surprised this hasn't been seen (or reported) previously. The "mypy_primer" [regression test](https://github.com/microsoft/pyright/pull/8621) that we run on every pyright PR indicates that this bug was affecting other code bases.

Thanks again for reporting the issue and taking the time to isolate it. Much appreciated!

This will be addressed in the next release.
